Add k8s trace attribute to PingSource

lionelvillard opened the pull request:
Partial fix for https://github.com/knative/eventing/issues/5888

Since OpenTelemetry does not support custom resources (yet), I chose what I think are reasonable span attribute names.

<!-- Please include the 'why' behind your changes if no issue exists -->

## Proposed Changes

<!-- Please categorize your changes:
- :gift: Add new feature
- :bug: Fix bug
- :broom: Update or clean up current behavior
- :wastebasket: Remove feature or internal logic
-->

- :gift: Traces generated by PingSource includes some k8s attributes  (`k8s.namespace`, `k8s.name`, `k8s.resource`).
-
-

### Pre-review Checklist

<!-- If these boxes are not checked, you will be asked to complete these requirements or explain why they do not apply to your PR. -->

- [x] **At least 80% unit test coverage**
- [x] (NA) **E2E tests** for any new behavior
- [x] (NA) **Docs PR** for any user-facing impact
- [x] (NA) **Spec PR** for any new API feature  
- [x] (NA) **Conformance test** for any change to the spec  

**Release Note**

<!--
:page_facing_up: If this change has user-visible impact, write a release note in the block
below. Include the string "action required" if additional action is required of
users switching to the new release, for example in case of a breaking change.

Write as if you are speaking to users, not other Knative contributors. If this
change has no user-visible impact, no release note is needed.
-->

```release-note
 :gift: Traces generated by PingSource includes some k8s attributes  (`k8s.namespace`, `k8s.name`, `k8s.resource`).
```


**Docs**

<!--
:book: If this change has user-visible impact, link to an issue or PR in
https://github.com/knative/docs.
-->

matzew commented:
what about other source?

odacremolbap commented on pkg/tracing/client/observability_service.go:
```diff
@@ -0,0 +1,79 @@
+		span.AddAttributes(spanData.Attributes...)
+	}
+
+	return ctx, func(errOrResult error) {
+		span.End()
+	}
```
The CE SDK implementation seems to bring a valuable tag to the span:

A result reporter is created per span:

https://github.com/cloudevents/sdk-go/blob/1170e89edb9b504a806f2c6a26563c3c26b68276/observability/opencensus/v2/client/observability_service.go#L44

And deferred to check the outcome, depending on it a tag is added:

https://github.com/cloudevents/sdk-go/blob/1170e89edb9b504a806f2c6a26563c3c26b68276/observability/opencensus/v2/client/observability_service.go#L50-L57

https://github.com/cloudevents/sdk-go/blob/1170e89edb9b504a806f2c6a26563c3c26b68276/observability/opencensus/v2/client/reporter.go#L77-L84

Do you think something like that could be added to this PR's implementation?

lionelvillard replied:
Thanks for bringing this up. The reason I removed the reporter is because it's about metrics, not traces, and I have on my TODO list to check if this is not redundant with [stats reporter](https://github.com/knative/eventing/blob/main/pkg/metrics/source/stats_reporter.go#L30)... and it's not. We need to have it for not breaking the existing behavior (though we don't document it)

odacremolbap commented on pkg/tracing/client/observability_service_test.go:
```diff
@@ -0,0 +1,66 @@
+/*
+Copyright 2020 The Knative Authors
```
```suggestion
Copyright 2021 The Knative Authors
```

odacremolbap commented on pkg/tracing/client/observability_service_test.go:
```diff
@@ -0,0 +1,66 @@
+	ctx := context.Background()
+	ctx = tracing.WithSpanData(ctx, "spanname", 1, []trace.Attribute{trace.StringAttribute("myattr", "myvalue")})
+
+	ctx, callback := New().RecordSendingEvent(ctx, event)
```
```suggestion
	_, callback := New().RecordSendingEvent(ctx, event)
```

odacremolbap reviewed (commented):
Added minor comments.
I think `result` tag would be very welcomed.

pierDipi commented on pkg/tracing/attributes.go:
```diff
@@ -28,6 +28,10 @@ const (
 	MessagingProtocolAttributeName    = "messaging.protocol"
 	MessagingMessageIDAttributeName   = "messaging.message_id"
+
+	K8sNamespaceName = "k8s.namespace"
+	K8sNameName      = "k8s.name"
+	K8sResourceName  = "k8s.resource" // Qualified resource name, eg. pingsource.sources.knative.dev
```
The [k8s semantic convention](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/k8s.md) uses:
`k8s.<resource>.name` for the name, so should we use a similar thing instead?

`k8s.pingsource.sources.knative.dev.name` (like `k8s.deployment.name`)

For the namespace, it should be:
`k8s.namespace.name`

lionelvillard replied:
> > Since OpenTelemetry does not support custom resources (yet), I chose what I think are reasonable span attribute names.
> 
> Should we propose one before we move too far down our custom names?

Yes I'm planning to. 

I was hesitant about using `k8s.<resource>.name` but it fits well into the current proposal, so +1. +1 as well for `k8s.namespace.name`. 

Thanks!

lionelvillard replied:
fixed.

pierDipi reviewed (commented):
> Since OpenTelemetry does not support custom resources (yet), I chose what I think are reasonable span attribute names.

Should we propose one before we move too far down our custom names?

lionelvillard commented:
> what about other source?

For core sources it is tracked in this issue: #5888. Does not apply to SinkBinding and ContainerSource as the adapter is external.

I can do APIServerSource in this PR.

lionelvillard commented:
> Added minor comments.
> I think `result` tag would be very welcomed.

Definitively, though maybe not in the PR as I'm not sure what should be the value. It could be as simple as ACK/NACK, but we may want something more/different. I feel this deserve its own issue, WDYT @odacremolbap ?

odacremolbap commented:
> > Added minor comments.
> > I think `result` tag would be very welcomed.
> 
> Definitively, though maybe not in the PR as I'm not sure what should be the value. It could be as simple as ACK/NACK, but we may want something more/different. I feel this deserve its own issue, WDYT @odacremolbap ?

sure, that can go to a different issue.

lionelvillard commented on pkg/tracing/client/observability_service.go:
```diff
@@ -0,0 +1,79 @@
+/*
+Copyright 2020 The Knative Authors
```
fixed

lionelvillard commented:
@odacremolbap I moved everything to the `observability` package and added the `result` tag to preserve the existing behavior. It's a bit messy right now but I would rather not invest too much time on this until we migrate to OpenTelemetry. At that point we can fold `stats_reporter` into the OpenTelemetry equivalent of the observer service introduced in this PR. 

@pierDipi The tracing semantic conventions for k8s are experimental anyway so I wouldn't wait. I'll update the doc to say these are experimental tracing attributes. Is that ok?

pierDipi reviewed (approved):
/lgtm

pierDipi commented:
/hold @odacremolbap review

pierDipi commented:
(too late, we can iterate in a follow up)

odacremolbap commented on pkg/adapter/v2/cloudevents.go:
```diff
@@ -23,19 +24,39 @@ import (
 )
 
-var newClientHTTPObserved = cloudeventsobsclient.NewClientHTTP
+var newClientHTTPObserved = NewClientHTTPObserved
```
nit: maybe `newClientHTTPObserved` is no longer needed and can be replaced with the public `MewClientHTTPObserved`

odacremolbap commented:
No worries @pierDipi , this also lgtm.
I have only one minor comment that can be addressed when we migrate to OpenTelemetry.

:+1:
